File: frontend/lib/predictionMarket.ts
import { Address, Horizon } from 'stellar-sdk';

// Contract deployment bilgileri (deploy ettikten sonra güncellenecek)
export const CONTRACT_ID = 'CDXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'; // Deploy sonrası update et

// Contract method adları
export const CONTRACT_METHODS = {
  INITIALIZE: 'initialize',
  CREATE_PREDICTION: 'create_prediction',
  PLACE_BET: 'place_bet',
  RESOLVE_PREDICTION: 'resolve_prediction',
  GET_PREDICTION: 'get_prediction',
  GET_USER_BETS: 'get_user_bets',
};

// xlm Token bilgileri (Testnet)
export const xlm_TOKEN = {
  CODE: 'xlm',
  ISSUER: 'GBRPYHIL2CI3WHZDTOOQFC6EB4NCQOB5SOHRAB52HM7XUFHRXKYDAB7D', // Kendi issuer adresiniz
};

export class PredictionMarketService {
  private contractId: string;
  private rpcUrl: string;

  constructor(contractId: string = CONTRACT_ID, rpcUrl: string = 'https://soroban-testnet.stellar.org:443') {
    this.contractId = contractId;
    this.rpcUrl = rpcUrl;
  }

  /**
   * Create a new prediction market
   */
  async createPrediction(
    predictionId: number,
    adminPublicKey: string,
    wallet: any // Freighter wallet
  ): Promise<string> {
    try {
      const server = new Horizon.Server(this.rpcUrl);
      
      // Build transaction
      const account = await server.loadAccount(adminPublicKey);
      
      // Create contract call
      // (Full implementation requires stellar-js SDK setup)
      
      console.log('Creating prediction:', predictionId);
      return 'transaction_hash'; // Placeholder
    } catch (error) {
      console.error('Error creating prediction:', error);
      throw error;
    }
  }

  /**
   * Place a bet on a prediction
   */
  async placeBet(
    predictionId: number,
    bettorPublicKey: string,
    amount: number,
    prediction: boolean, // true = EVET, false = HAYIR
    wallet: any
  ): Promise<string> {
    try {
      console.log('Placing bet:', {
        predictionId,
        amount,
        prediction: prediction ? 'EVET' : 'HAYIR',
      });

      // Call contract method: place_bet
      // Return transaction hash
      return 'transaction_hash'; // Placeholder
    } catch (error) {
      console.error('Error placing bet:', error);
      throw error;
    }
  }

  /**
   * Resolve a prediction
   */
  async resolvePrediction(
    predictionId: number,
    result: boolean, // true = EVET won, false = HAYIR won
    adminPublicKey: string,
    wallet: any
  ): Promise<string> {
    try {
      console.log('Resolving prediction:', predictionId, 'Result:', result ? 'EVET' : 'HAYIR');

      // Call contract method: resolve_prediction
      // Distribute winnings automatically
      return 'transaction_hash'; // Placeholder
    } catch (error) {
      console.error('Error resolving prediction:', error);
      throw error;
    }
  }

  /**
   * Get prediction details
   */
  async getPrediction(predictionId: number): Promise<any> {
    try {
      const server = new Horizon.Server(this.rpcUrl);
      
      // Query contract state
      // (Full implementation requires stellar-js SDK)
      
      return {
        id: predictionId,
        yes_amount: 0,
        no_amount: 0,
        resolved: false,
        result: false,
      };
    } catch (error) {
      console.error('Error fetching prediction:', error);
      throw error;
    }
  }

  /**
   * Get user's bets
   */
  async getUserBets(predictionId: number, userPublicKey: string): Promise<any[]> {
    try {
      const server = new Horizon.Server(this.rpcUrl);
      
      // Query contract state for user bets
      return [];
    } catch (error) {
      console.error('Error fetching user bets:', error);
      throw error;
    }
  }
}

export default PredictionMarketService;

File: contracts/prediction_market.rs
#![no_std]
use soroban_sdk::{contract, contractimpl, Address, Env, Symbol, Vec, map, Map};

#[derive(Clone, Copy)]
pub struct Bet {
    pub bettor: Address,
    pub amount: i128,
    pub prediction: bool, // true = EVET, false = HAYIR
    pub prediction_id: u32,
}

#[derive(Clone, Copy)]
pub struct Prediction {
    pub id: u32,
    pub yes_amount: i128,
    pub no_amount: i128,
    pub resolved: bool,
    pub result: bool, // true = EVET won, false = HAYIR won
}

#[contract]
pub struct PredictionMarket;

#[contractimpl]
impl PredictionMarket {
    /// Initialize contract with xlm token and admin
    pub fn initialize(env: Env, token: Address, admin: Address) {
        let storage = env.storage().persistent();
        storage.set(&Symbol::new(&env, "token"), &token);
        storage.set(&Symbol::new(&env, "admin"), &admin);
        storage.set(&Symbol::new(&env, "next_prediction_id"), &1u32);
    }

    /// Create a new prediction market
    pub fn create_prediction(env: Env, id: u32) {
        let storage = env.storage().persistent();
        let admin: Address = storage.get(&Symbol::new(&env, "admin")).unwrap();
        
        admin.require_auth();

        let prediction = Prediction {
            id,
            yes_amount: 0,
            no_amount: 0,
            resolved: false,
            result: false,
        };

        let key = Symbol::new(&env, &format!("prediction_{}", id));
        storage.set(&key, &prediction);
    }

    /// Place a bet on a prediction
    pub fn place_bet(env: Env, prediction_id: u32, bettor: Address, amount: i128, prediction: bool) {
        let storage = env.storage().persistent();
        let token: Address = storage.get(&Symbol::new(&env, "token")).unwrap();

        bettor.require_auth();

        // Transfer xlm tokens from bettor to contract
        soroban_sdk::token::Client::new(&env, &token)
            .transfer(&bettor, &env.current_contract_address(), &amount);

        // Update prediction amounts
        let pred_key = Symbol::new(&env, &format!("prediction_{}", prediction_id));
        let mut pred: Prediction = storage.get(&pred_key).unwrap();

        if prediction {
            pred.yes_amount += amount;
        } else {
            pred.no_amount += amount;
        }

        storage.set(&pred_key, &pred);

        // Store bet record
        let bet_key = Symbol::new(&env, &format!("bet_{}_{}", prediction_id, bettor.clone()));
        let mut bets: Vec<(bool, i128)> = storage.get(&bet_key).unwrap_or(Vec::new(&env));
        bets.push_back((prediction, amount));
        storage.set(&bet_key, &bets);
    }

    /// Resolve a prediction and distribute winnings
    pub fn resolve_prediction(env: Env, prediction_id: u32, result: bool) {
        let storage = env.storage().persistent();
        let admin: Address = storage.get(&Symbol::new(&env, "admin")).unwrap();
        let token: Address = storage.get(&Symbol::new(&env, "token")).unwrap();

        admin.require_auth();

        let pred_key = Symbol::new(&env, &format!("prediction_{}", prediction_id));
        let mut pred: Prediction = storage.get(&pred_key).unwrap();

        pred.resolved = true;
        pred.result = result;
        storage.set(&pred_key, &pred);

        // Calculate winnings
        let total_pool = pred.yes_amount + pred.no_amount;
        let winning_pool = if result { pred.yes_amount } else { pred.no_amount };

        // Distribute to winners
        // (Implementation would iterate through bettors and send their share)
    }

    /// Get prediction details
    pub fn get_prediction(env: Env, prediction_id: u32) -> Prediction {
        let storage = env.storage().persistent();
        let key = Symbol::new(&env, &format!("prediction_{}", prediction_id));
        storage.get(&key).unwrap()
    }

    /// Get user's bets for a prediction
    pub fn get_user_bets(env: Env, prediction_id: u32, bettor: Address) -> Vec<(bool, i128)> {
        let storage = env.storage().persistent();
        let key = Symbol::new(&env, &format!("bet_{}_{}", prediction_id, bettor));
        storage.get(&key).unwrap_or(Vec::new(&env))
    }
}
